### PR TITLE
feat(ui): option for automated update checks

### PR DESCRIPTION
**Pull request by IndusAryan**
this new update checker is all about making it easier to find new updates, immediately when they are available as well as now users dont have to check them manually, idk if it has to do with fdroid issue but most people use fdroid for sake of auto update imo, this dialog also redirrects to latest release tag when download is clicked. now users can also see changelog in app itself that are cleaned through regex sanitization.

OLD :

photo_2024-01-20_17-13-42

NEW ->

photo_1_2024-02-27_17-22-51

photo_2_2024-02-27_17-22-51

### CODE REVIEW

**Comment by akano12** [comment]
Ye expect people to complain about this feature and ask for an option to disable auto updates (assuming this will popup on each app launch when there's an update)

**Comment by IndusAryan** [comment]
whats the sense of not updating ? just press download and retrieve 20mb asset , install and gtg. it would make sense if our app was recieving updates daily, or was above 100mb, even for 1 update once a month.
people can cope hard.

**Comment by Bnyro** [comment]
We already have koltinx serialization, it doesn't make sense to add gson.

**Comment by Bnyro** [comment]
And it should be disabled by default, because most users don't like to auto-connect to GitHub and expose their IP by that.

**Comment by IndusAryan** [comment]
> And it should be disabled by default, because most users don't like to auto-connect to GitHub and expose their IP by that.

instead of new preference, i think I should connect it with proxy preference? are you down? because manual button is already there.

**Comment by Bnyro** [comment]
> instead of new preference, i think I should connect it with proxy preference? are you down? because manual button is already there.
>
No, it should be a new preference in the advanced settings.

File: app/src/main/java/com/github/libretube/util/InAppUpdater.kt [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    val body: String, // changelog
+    val html_url: String // uri to latest release tag
+)
+
+interface GitHubService {
+    @GET("repos/libre-tube/LibreTube/releases/latest")
+    suspend fun getLatestRelease(): Response<Release>
+}
+
+class InAppUpdater(private val context: Context) {
```
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/util/InAppUpdater.kt:28]:
```suggestion
class UpdateChecker(private val context: Context) {
```

File: app/src/main/java/com/github/libretube/util/InAppUpdater.kt [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+data class Release(
+    val name: String, // version name
+    val body: String, // changelog
+    val html_url: String // uri to latest release tag
+)
+
+interface GitHubService {
+    @GET("repos/libre-tube/LibreTube/releases/latest")
+    suspend fun getLatestRelease(): Response<Release>
+}
```
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/util/InAppUpdater.kt:26]: Please keep that in separate files

File: app/src/main/java/com/github/libretube/util/InAppUpdater.kt [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    private lateinit var changelog: String
+    private lateinit var releaseURL: String
+
+    suspend fun checkUpdate(manualTrigger: Boolean = false) {
+
+        val currentAppVersion = BuildConfig.VERSION_NAME.replace(".","")
+
+        try {
+            val retrofit = Retrofit.Builder()
+                .baseUrl("https://api.github.com/")
```
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/util/InAppUpdater.kt:39]: move to a constant as before

File: app/src/main/java/com/github/libretube/util/InAppUpdater.kt [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+                        UpdateAvailableDialog().showDialog(sanitizeChangelog(changelog),releaseURL, context)
+                    }
+
+                } else if (manualTrigger){
+                    Toast.makeText(context, R.string.app_uptodate, Toast.LENGTH_LONG).show()
+                }
+
+            } else {
+                Toast.makeText(context,R.string.unknown_error, Toast.LENGTH_LONG).show()
+                Log.e(TAG(),"$response")
```
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/util/InAppUpdater.kt:65]:
```suggestion
                Log.e(TAG(), response)
```

File: app/src/main/java/com/github/libretube/util/InAppUpdater.kt [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        } catch (e: Exception) {
+            e.printStackTrace()
+        }
+    }
+
+    private fun sanitizeChangelog(changelog: String): String {
+        // because it is useless
+        val sanitizedChangelog = changelog.substringBeforeLast("**Full Changelog**")
+
+        // links are ugly af
```
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/util/InAppUpdater.kt:77]: no weird comments needed

File: app/src/main/java/com/github/libretube/util/InAppUpdater.kt (hunk 1 above)
Comment (IndusAryan) [reply app/src/main/java/com/github/libretube/util/InAppUpdater.kt]: done this

File: app/src/main/java/com/github/libretube/util/InAppUpdater.kt (hunk 2 above)
Comment (IndusAryan) [reply app/src/main/java/com/github/libretube/util/InAppUpdater.kt]: moved them to external api

File: app/src/main/java/com/github/libretube/util/InAppUpdater.kt (hunk 4 above)
Comment (IndusAryan) [reply app/src/main/java/com/github/libretube/util/InAppUpdater.kt]: ok

File: app/src/main/java/com/github/libretube/util/InAppUpdater.kt (hunk 5 above)
Comment (IndusAryan) [reply app/src/main/java/com/github/libretube/util/InAppUpdater.kt]: removed, dont mind them please, ik i lack professionalism

**Comment by IndusAryan** [comment]
> We already have koltinx serialization, it doesn't make sense to add gson.

> No, it should be a new preference in the advanced settings.

removed gson, now it uses serialization and retrofit instance instead of builder.

File: app/src/main/java/com/github/libretube/util/InAppUpdater.kt (hunk 3 above)
Comment (IndusAryan) [reply app/src/main/java/com/github/libretube/util/InAppUpdater.kt]: ya

**Comment by Bnyro** [comment]
> I like my apps to be updated from one source and that's neo store, it has already background updates, so it would be a backstep in comfort. Currently I get my updates immediately if they are released in izzydroid repo without doing anything (not even opening libretube).
>
That's not changed here, the title and description of this PR were just very misleading. In fact, this only checks for new updates automatically (and is disabled by default). We already had an in-app updater once and I removed it for a reason, something similar won't be added again. That's what app stores exist for.

File: app/src/main/java/com/github/libretube/ui/dialogs/UpdateAvailableDialog.kt [hunk 6]
```diff
@@ -1,33 +1,28 @@
...
+            .setPositiveButton(R.string.download) { _, _ ->
+                val intent = Intent(Intent.ACTION_VIEW).setData(releaseUrl.toUri())
+                startActivity(context,intent, Bundle())
+            }
+            .setNegativeButton(R.string.tooltip_dismiss) { dialog, _ ->
+                dialog.dismiss()
             }
-            .show()
+
+        dialogBuilder.show()
```
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/ui/dialogs/UpdateAvailableDialog.kt:26]: It's useless to save the dialog builder to a variable, `.show()` should be done right below `setNegativeButton`

File: app/src/main/java/com/github/libretube/api/ExternalApi.kt
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/api/ExternalApi.kt:1]: There's no reason for any changes to this file...

File: app/src/main/java/com/github/libretube/obj/update/UpdateInfo.kt
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/obj/update/UpdateInfo.kt:1]: There's no reason for any changes to this file...

File: app/src/main/java/com/github/libretube/ui/activities/MainActivity.kt [hunk 7]
```diff
@@ -81,6 +84,14 @@ class MainActivity : BaseActivity() {
         binding = ActivityMainBinding.inflate(layoutInflater)
         setContentView(binding.root)
 
+        if (PreferenceHelper.getBoolean(PreferenceKeys.AUTO_UPDATES, false)) {
+
```
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/ui/activities/MainActivity.kt:88]: useless blank line

File: app/src/main/java/com/github/libretube/util/InAppUpdater.kt [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+    private lateinit var changelog: String
+    private lateinit var releaseURL: String
+
+    suspend fun checkUpdate(manualTrigger: Boolean = false) {
+
+        val currentAppVersion = BuildConfig.VERSION_NAME.replace(".", "")
+
+        try {
+
```
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/util/InAppUpdater.kt:24]: useless blank line

File: app/src/main/java/com/github/libretube/util/InAppUpdater.kt [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+
+class UpdateChecker(private val context: Context) {
+
+    private lateinit var changelog: String
+    private lateinit var releaseURL: String
+
+    suspend fun checkUpdate(manualTrigger: Boolean = false) {
+
```
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/util/InAppUpdater.kt:20]: useless blank line

File: app/src/main/java/com/github/libretube/util/InAppUpdater.kt [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+                if (update != null && currentAppVersion.toInt() < update) {
+                    withContext(Dispatchers.Main) {
+                        UpdateAvailableDialog().showDialog(
+                            sanitizeChangelog(changelog),
+                            releaseURL,
+                            context
+                        )
+                    }
+
```
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/util/InAppUpdater.kt:43]: useless blank line

File: app/src/main/java/com/github/libretube/util/InAppUpdater.kt [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+                } else if (manualTrigger) {
+                    Toast.makeText(context, R.string.app_uptodate, Toast.LENGTH_LONG).show()
+                }
+
```
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/util/InAppUpdater.kt:47]: useless blank line

File: app/src/main/java/com/github/libretube/util/InAppUpdater.kt [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+            } else {
+                Toast.makeText(context, R.string.unknown_error, Toast.LENGTH_LONG).show()
+                Log.e(TAG(), response.toString())
+            }
+
```
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/util/InAppUpdater.kt:52]: useless blank line

File: app/src/main/java/com/github/libretube/util/InAppUpdater.kt [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+        } catch (e: Exception) {
+            e.printStackTrace()
+        }
+    }
+
+    private fun sanitizeChangelog(changelog: String): String {
+
```
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/util/InAppUpdater.kt:59]: useless blank line

File: app/src/main/java/com/github/libretube/obj/update/UpdateInfo.kt
Comment (IndusAryan) [reply app/src/main/java/com/github/libretube/obj/update/UpdateInfo.kt:1]: wdym, why it is needed now if its not used

File: app/src/main/java/com/github/libretube/api/ExternalApi.kt
Comment (IndusAryan) [reply app/src/main/java/com/github/libretube/api/ExternalApi.kt:1]: you asked to movie data class and get interface to a seperate file?!

File: app/src/main/java/com/github/libretube/ui/preferences/MainSettings.kt [hunk 14]
```diff
@@ -24,50 +16,18 @@ class MainSettings : BasePreferenceFragment() {
...
-                        bundleOf(IntentData.updateInfo to updateInfo)
-                    newUpdateAvailableDialog.show(
-                        childFragmentManager,
-                        UpdateAvailableDialog::class.java.name
-                    )
-                } else {
-                    // otherwise show the no update available snackBar
-                    showSnackBar(R.string.app_uptodate)
-                }
+            lifecycleScope.launch {
```
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/ui/preferences/MainSettings.kt:25]:
```suggestion
            lifecycleScope.launch(Dispatchers.IO) {
```

**Comment by IndusAryan** [comment]
> That's not changed here, the title and description of this PR were just very misleading. In fact, this only checks for new updates automatically (and is disabled by default). We already had an in-app updater once and I removed it for a reason, something similar won't be added again. That's what app stores exist for.

which updater you are talking about that will never come? the old updater was static was static with 2 button and nil information. are you talking about one which download apk when pressed ok and prompts intall? this pr was named so because thats what i am gonna do, but if its out of scope, i won't now.

File: app/src/main/java/com/github/libretube/ui/preferences/MainSettings.kt (hunk 14 above)
Comment (IndusAryan) [reply app/src/main/java/com/github/libretube/ui/preferences/MainSettings.kt]: ok

**Comment by Bnyro** [comment]
> which download apk when pressed ok and prompts intall
>
Yes, that's what I'm talking about. We already had that but removed it as it's nothing an app should do.

**Comment by IndusAryan** [comment]
bnyro , can you run lint workflo again here? i will fix the lint after looking at log

**Comment by itszechs** [comment]
haven't looked at latest version of this PR, but from the initial screenshots I'd suggest parsing the changelog https://github.com/noties/Markwon

**Comment by IndusAryan** [comment]
> haven't looked at latest version of this PR, but from the initial screenshots I'd suggest parsing the changelog https://github.com/noties/Markwon

its not needed afaik, the changelog is already verbose and clean with relevant info only.

**Comment by IndusAryan** [comment]
> > which download apk when pressed ok and prompts intall
> 
> Yes, that's what I'm talking about. We already had that but removed it as it's nothing an app should do.

i have made some final changes that are,
made api responses changelog and releaseURl in update checker vars nullable (in case they are not responded by api)
made dialog an extension of dialog fragment. 

i think its good for your final review

File: app/src/main/java/com/github/libretube/api/ExternalApi.kt [hunk 15]
```diff
@@ -18,7 +19,7 @@ interface ExternalApi {
 
     // fetch latest version info
     @GET(GITHUB_API_URL)
-    suspend fun getUpdateInfo(): UpdateInfo
+    suspend fun getLatestRelease(): Response<UpdateInfo>
```
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/api/ExternalApi.kt:22]: There's no reason to change the response type to `Response<UpdateInfo>`

File: app/src/main/java/com/github/libretube/ui/activities/MainActivity.kt [hunk 16]
```diff
@@ -82,6 +86,13 @@ class MainActivity : BaseActivity() {
         binding = ActivityMainBinding.inflate(layoutInflater)
         setContentView(binding.root)
 
+        if (PreferenceHelper.getBoolean(PreferenceKeys.AUTO_UPDATES, false)) {
```
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/ui/activities/MainActivity.kt:89]: The preference key should be called `AUTOMATIC_UPDATE_CHECKS`

File: app/src/main/java/com/github/libretube/ui/preferences/MainSettings.kt [hunk 17]
```diff
@@ -24,50 +17,18 @@ class MainSettings : BasePreferenceFragment() {
         setPreferencesFromResource(R.xml.settings, rootKey)
 
         val update = findPreference<Preference>("update")
+        val appVersion = BuildConfig.VERSION_NAME
+        update?.summary = "v$appVersion"
```
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/ui/preferences/MainSettings.kt:21]:
```suggestion
        update?.summary = "v${BuildConfig.VERSION_NAME}"
```

File: app/src/main/res/xml/settings.xml [hunk 18]
```diff
@@ -75,11 +75,10 @@
     <PreferenceCategory>
 
         <Preference
-            android:icon="@drawable/ic_update"
+            android:icon="@drawable/ic_download"
```
Comment (Bnyro) [inline app/src/main/res/xml/settings.xml:78]: Why?

File: app/src/main/res/xml/advanced_settings.xml [hunk 19]
```diff
@@ -4,6 +4,12 @@
 
     <PreferenceCategory app:title="@string/advanced">
 
+        <SwitchPreferenceCompat
+            android:defaultValue="false"
+            android:icon="@drawable/ic_download_filled"
+            app:key="auto_update"
```
Comment (Bnyro) [inline app/src/main/res/xml/advanced_settings.xml:10]:
```suggestion
            app:key="automatic_update_checks"
```

File: app/src/main/res/values/strings.xml [hunk 20]
```diff
@@ -420,6 +420,7 @@
     <string name="group_name_error_exists">Please choose a name that is unique</string>
     <string name="edit_group">Edit group</string>
     <string name="play_automatically">Play automatically</string>
+    <string name="show_updates">Show app updates</string>
```
Comment (Bnyro) [inline app/src/main/res/values/strings.xml:423]:
```suggestion
    <string name="show_updates">Automatically check for new updates</string>
```

File: app/src/main/res/values/strings.xml [hunk 21]
```diff
@@ -155,7 +155,7 @@
     <string name="empty_instance">Fill in the name and the API URL.</string>
     <string name="clear_customInstances">Clear added</string>
     <string name="invalid_url">Please enter a URL that works</string>
-    <string name="version">Version %1$s</string>
+    <string name="version">v0.21.1</string>
```
Comment (Bnyro) [inline app/src/main/res/values/strings.xml:158]: Undo

File: app/src/main/java/com/github/libretube/util/UpdateChecker.kt [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+import com.github.libretube.R
+import com.github.libretube.api.RetrofitInstance
+import com.github.libretube.extensions.TAG
+import com.github.libretube.ui.dialogs.UpdateAvailableDialog
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+
+class UpdateChecker(private val context: Context) {
+    private var changelog: String? = null
+    private var releaseURL: String? = null
```
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/util/UpdateChecker.kt:16]: These variables should only be declared in the `checkUpdate` method.

File: app/src/main/java/com/github/libretube/util/UpdateChecker.kt [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+            }
+        } catch (e: Exception) {
+            e.printStackTrace()
+        }
+    }
+
+    private fun sanitizeChangelog(changelog: String): String {
+        val sanitizedChangelog = changelog.substringBeforeLast("**Full Changelog**")
+        val withoutLinks = sanitizedChangelog.replace(Regex("in https://github\\.com/[^\\s]+"), "")
+
```
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/util/UpdateChecker.kt:55]: `## ` for titles should also be removed here in my opinion.

Comment (IndusAryan) [reply app/src/main/java/com/github/libretube/util/UpdateChecker.kt:55]: 👍🏽 removed plus replaced the asterisks with small clean bullet points, image updated in pr description

File: app/src/main/res/values/strings.xml (hunk 20 above)
Comment (IndusAryan) [reply app/src/main/res/values/strings.xml]: renamed to check updates automatically as the title was large and hidden in settings

File: app/src/main/res/xml/settings.xml (hunk 18 above)
Comment (IndusAryan) [reply app/src/main/res/xml/settings.xml:78]: it is so to make the drawable differ from manual update check button, no drawable has been added, it is existing.

File: app/src/main/java/com/github/libretube/util/UpdateChecker.kt [hunk 24]
```diff
@@ -10,49 +10,49 @@ import com.github.libretube.extensions.TAG
...
-                    withContext(Dispatchers.Main) {
-                        UpdateAvailableDialog().onCreateDialog(
-                            sanitizeChangelog(changelog!!),
-                            releaseURL!!,
-                            context,
-                        )
-                    }
-                } else if (manualTrigger) {
-                    Toast.makeText(context, R.string.app_uptodate, Toast.LENGTH_LONG).show()
+            releaseURL = response.htmlUrl
```
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/util/UpdateChecker.kt:23]:
```suggestion
            val releaseURL = response.htmlUrl
```

File: app/src/main/java/com/github/libretube/util/UpdateChecker.kt [hunk 25]
```diff
@@ -10,49 +10,49 @@ import com.github.libretube.extensions.TAG
...
+
+            // version would be in the format "0.21.1"
+            val update = response.name.replace(".", "").toIntOrNull()
+            changelog = response.body
+
+            if (update != null && currentAppVersion < update) {
+                withContext(Dispatchers.Main) {
+                    UpdateAvailableDialog().onCreateDialog(
+                        sanitizeChangelog(changelog),
```
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/util/UpdateChecker.kt:32]:
```suggestion
                        sanitizeChangelog(response.body),
```

File: app/src/main/java/com/github/libretube/util/UpdateChecker.kt [hunk 26]
```diff
@@ -10,49 +10,49 @@ import com.github.libretube.extensions.TAG
...
+                        releaseURL,
```
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/util/UpdateChecker.kt:33]:
```suggestion
                        response.htmlURL,
```

File: app/src/main/java/com/github/libretube/util/UpdateChecker.kt [hunk 27]
```diff
@@ -10,49 +10,49 @@ import com.github.libretube.extensions.TAG
...
 
 class UpdateChecker(private val context: Context) {
-    private var changelog: String? = null
-    private var releaseURL: String? = null
 
-    suspend fun checkUpdate(manualTrigger: Boolean = false) {
-        val currentAppVersion = BuildConfig.VERSION_NAME.replace(".", "")
+    suspend fun checkUpdate(isManualCheck: Boolean = false) {
+        val changelog: String?
+        val releaseURL: String?
```
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/util/UpdateChecker.kt:18]:
```suggestion
```

File: app/src/main/java/com/github/libretube/util/UpdateChecker.kt [hunk 28]
```diff
@@ -10,49 +10,49 @@ import com.github.libretube.extensions.TAG
...
+                        releaseURL,
+                        context,
+                    )
                 }
-            } else {
-                Toast.makeText(context, R.string.unknown_error, Toast.LENGTH_LONG).show()
-                Log.e(TAG(), response.toString())
+                Log.i(TAG(), response.toString())
+            } else if (isManualCheck) {
+                Toast.makeText(context, R.string.app_uptodate, Toast.LENGTH_LONG).show()
```
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/util/UpdateChecker.kt:39]:
```suggestion
                context.toastFromMainDispatcher(R.string.app_uptodate)
```

File: app/src/main/java/com/github/libretube/util/UpdateChecker.kt [hunk 29]
```diff
@@ -10,49 +10,49 @@ import com.github.libretube.extensions.TAG
...
-                            context,
-                        )
-                    }
-                } else if (manualTrigger) {
-                    Toast.makeText(context, R.string.app_uptodate, Toast.LENGTH_LONG).show()
+            releaseURL = response.htmlUrl
+
+            // version would be in the format "0.21.1"
+            val update = response.name.replace(".", "").toIntOrNull()
+            changelog = response.body
```
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/util/UpdateChecker.kt:27]:
```suggestion
            // version would be in the format "0.21.1"
            val update = response.name.replace(".", "").toIntOrNull()
```

**Comment by IndusAryan** [comment]
ohh, god, bnyro how do you fix ktlint problems locally, they are asking to install pinterest dependency for windows?

**Pull request closed without merging**